content/chapters/data/lab/content/investigate-memory.md

```diff
@@ -0,0 +1,510 @@
+==24671==
+==24671== LEAK SUMMARY:
+==24671==    definitely lost: 0 bytes in 0 blocks
+==24671==    indirectly lost: 0 bytes in 0 blocks
+==24671==      possibly lost: 0 bytes in 0 blocks
+==24671==    still reachable: 264,929 bytes in 25 blocks
+==24671==         suppressed: 0 bytes in 0 blocks
+==24671== Rerun with --leak-check=full to see details of leaked memory
+==24671==
+==24671== For counts of detected and suppressed errors, rerun with: -v
+==24671== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
+
+student@os:~/.../lab/support/memory-leak$ valgrind bash -c 'echo "ha"'
+==24675== Memcheck, a memory error detector
+==24675== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
+==24675== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
+==24675== Command: bash -c echo\ "ha"
+==24675==
+ha
+==24675==
+==24675== HEAP SUMMARY:
+==24675==     in use at exit: 43,056 bytes in 672 blocks
+==24675==   total heap usage: 774 allocs, 102 frees, 51,405 bytes allocated
+==24675==
+==24675== LEAK SUMMARY:
+==24675==    definitely lost: 12 bytes in 1 blocks
+==24675==    indirectly lost: 0 bytes in 0 blocks
+==24675==      possibly lost: 0 bytes in 0 blocks
+==24675==    still reachable: 43,044 bytes in 671 blocks
+==24675==         suppressed: 0 bytes in 0 blocks
+==24675== Rerun with --leak-check=full to see details of leaked memory
+==24675==
+==24675== For counts of detected and suppressed errors, rerun with: -v
+==24675== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
+```
+
+We can see that `ls` and `ps` don't have memory leaks.
+However, the shell (Bash) shows a memory leak of 12 bytes (on the test system).
+This may be a false positive or the subject of an actual investigation.
+
+Note that the `still reachable` section of the output refers to memory that wasn't freed but still has pointers referring it.
+A true memory leak occurs when no pointers refer any memory area.
+
+#### Practice
+
+1. Investigate 2-3 other executables in the system using Valgrind.
+
+1. Use `ltrace` to list `malloc()` and `free()` calls made by the investigated system executables.
+
+#### Quiz
+
+TODO
+
+### jemalloc
+
+[jemalloc](http://jemalloc.net/) is a featureful allocator that is intended to replace the standard allocator in the standard C library (libc).
+jemalloc provides replacements for the general `malloc()` and `free()` functions and also provides a custom API targeted for performance tuning.
+
+As [documented](https://github.com/jemalloc/jemalloc/wiki/Getting-Started) there are multiple ways to use `jemalloc`, the easiest of which is to use the `LD_PRELOAD` environment variable and preload the library and hook into `malloc()` and `free()` function calls.
+
+First install `jemalloc` on our system.
+On your typical Ubuntu / Debian-based system, use `apt`:
+
+```
+student@os:~/.../data/lab/content$ sudo apt -y install libjemalloc-dev libjemalloc1
```

Comment:
libjemalloc1 may not be present on some systems. it appears to be called libjemalloc2 on some. libjemalloc-dev should provide libjemalloc[12] as a dependency.